File: hu/codes/java/chapter_stack_and_queue/linkedlist_queue.java
/**
 * File: linkedlist_queue.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_stack_and_queue;

import java.util.*;

/* Láncolt lista alapú sor implementáció */
class LinkedListQueue {
    private ListNode front, rear; // Fejcsomópont front, farokcsomópont rear
    private int queSize = 0;

    public LinkedListQueue() {
        front = null;
        rear = null;
    }

    /* Sor hosszának lekérése */
    public int size() {
        return queSize;
    }

    /* Sor üresség ellenőrzése */
    public boolean isEmpty() {
        return size() == 0;
    }

    /* Elem sorba rakása */
    public void push(int num) {
        // num hozzáadása a farokcsomópont után
        ListNode node = new ListNode(num);
        // Ha a sor üres, front és rear is a csomópontra mutat
        if (front == null) {
            front = node;
            rear = node;
        // Ha a sor nem üres, a csomópontot a farokcsomópont után adja hozzá
        } else {
            rear.next = node;
            rear = node;
        }
        queSize++;
    }

    /* Elem kivétele a sorból */
    public int pop() {
        int num = peek();
        // Fejcsomópont törlése
        front = front.next;
        queSize--;
        return num;
    }

    /* Lista visszaadása nyomtatáshoz */
    public int peek() {
        if (isEmpty())
            throw new IndexOutOfBoundsException();
        return front.val;
    }

    /* Láncolt lista tömbbé konvertálása és visszaadása */
    public int[] toArray() {
        ListNode node = front;
        int[] res = new int[size()];
        for (int i = 0; i < res.length; i++) {
            res[i] = node.val;
            node = node.next;
        }
        return res;
    }
}

public class linkedlist_queue {
    public static void main(String[] args) {
        /* Sor elejének elérése */
        LinkedListQueue queue = new LinkedListQueue();

        /* Elemek sorba rakása */
        queue.push(1);
        queue.push(3);
        queue.push(2);
        queue.push(5);
        queue.push(4);
        System.out.println("Queue queue = " + Arrays.toString(queue.toArray()));

        /* Lista visszaadása nyomtatáshoz */
        int peek = queue.peek();
        System.out.println("Front element peek = " + peek);

        /* Elem kivétele a sorból */
        int pop = queue.pop();
        System.out.println("Dequeue element pop = " + pop + ", after dequeue, queue = " + Arrays.toString(queue.toArray()));

        /* Sor hosszának lekérése */
        int size = queue.size();
        System.out.println("Queue length size = " + size);

        /* Sor üresség ellenőrzése */
        boolean isEmpty = queue.isEmpty();
        System.out.println("Queue is empty = " + isEmpty);
    }
}

File: hu/codes/python/chapter_stack_and_queue/linkedlist_queue.py
"""
File: linkedlist_queue.py
Created Time: 2022-12-01
Author: Peng Chen (pengchzn@gmail.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import ListNode


class LinkedListQueue:
    """Láncolt listán alapuló sor implementáció"""

    def __init__(self):
        """Konstruktor"""
        self._front: ListNode | None = None  # Fejcsomópont (sor eleje)
        self._rear: ListNode | None = None  # Farokcsomópont (sor vége)
        self._size: int = 0

    def size(self) -> int:
        """A sor hosszának lekérése"""
        return self._size

    def is_empty(self) -> bool:
        """Ellenőrzi, hogy a sor üres-e"""
        return self._size == 0

    def push(self, num: int):
        """Sorba állás (enqueue)"""
        # A num értéket a farokcsomópont után adja hozzá
        node = ListNode(num)
        # Ha a sor üres, mind az elülső, mind a hátsó pointer a csomópontra mutat
        if self._front is None:
            self._front = node
            self._rear = node
        # Ha a sor nem üres, a csomópontot a farokcsomópont után adja hozzá
        else:
            self._rear.next = node
            self._rear = node
        self._size += 1

    def pop(self) -> int:
        """Sorból kivétel (dequeue)"""
        num = self.peek()
        # Fejcsomópont törlése
        self._front = self._front.next
        self._size -= 1
        return num

    def peek(self) -> int:
        """A sor elejét adja vissza"""
        if self.is_empty():
            raise IndexError("A sor üres")
        return self._front.val

    def to_list(self) -> list[int]:
        """Lista formájú konverzió nyomtatáshoz"""
        queue = []
        temp = self._front
        while temp:
            queue.append(temp.val)
            temp = temp.next
        return queue


"""Fő kód"""
if __name__ == "__main__":
    # Sor inicializálása
    queue = LinkedListQueue()

    # Elemek sorba állítása
    queue.push(1)
    queue.push(3)
    queue.push(2)
    queue.push(5)
    queue.push(4)
    print("sor =", queue.to_list())

    # A sor elején lévő elem elérése
    peek: int = queue.peek()
    print("Sor eleje front =", peek)

    # Elem kivétele a sorból
    pop_front: int = queue.pop()
    print("Kivett elem pop =", pop_front)
    print("Kivétel után sor =", queue.to_list())

    # A sor hosszának lekérése
    size: int = queue.size()
    print("Sor hossza size =", size)

    # Ellenőrzi, hogy a sor üres-e
    is_empty: bool = queue.is_empty()
    print("A sor üres-e =", is_empty)
